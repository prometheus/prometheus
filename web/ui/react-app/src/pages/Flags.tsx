import React, { FC } from 'react';
import { RouteComponentProps } from '@reach/router';
import { Alert, Table } from 'reactstrap';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faSpinner } from '@fortawesome/free-solid-svg-icons';
import { useFetch } from '../utils/useFetch';
import PathPrefixProps from '../PathPrefixProps';

export interface FlagMap {
  [key: string]: string;
}

const Flags: FC<RouteComponentProps & PathPrefixProps> = ({ pathPrefix }) => {
  const { response, error } = useFetch(`${pathPrefix}/api/v1/status/flags`);

  const body = () => {
    const flags: FlagMap = response && response.data;
    if (error) {
      return (
        <Alert color="danger">
          <strong>Error:</strong> Error fetching flags: {error.message}
        </Alert>
      );
    } else if (flags) {
      return (
        <Table bordered={true} size="sm" striped={true}>
          <tbody>
            {Object.keys(flags).map(key => {
              return (
                <tr key={key}>
                  <th>{key}</th>
                  <td>{flags[key]}</td>
                </tr>
              );
            })}
          </tbody>
        </Table>
      );
    }
    return <FontAwesomeIcon icon={faSpinner} spin />;
  };

  return (
    <>
      <h2>Command-Line Flags</h2>
      {body()}
    </>
  );
};

export default Flags;
