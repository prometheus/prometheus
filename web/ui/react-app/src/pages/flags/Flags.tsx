import React, { FC } from 'react';
import { RouteComponentProps } from '@reach/router';
import { Table } from 'reactstrap';
import { withStatusIndicator } from '../../components/withStatusIndicator';
import { useFetch } from '../../hooks/useFetch';
import { usePathPrefix } from '../../contexts/PathPrefixContext';
import { API_PATH } from '../../constants/constants';

interface FlagMap {
  [key: string]: string;
}

interface FlagsProps {
  data?: FlagMap;
}

export const FlagsContent: FC<FlagsProps> = ({ data = {} }) => {
  return (
    <>
      <h2>Command-Line Flags</h2>
      <Table bordered size="sm" striped>
        <tbody>
          {Object.keys(data).map(key => (
            <tr key={key}>
              <th>{key}</th>
              <td>{data[key]}</td>
            </tr>
          ))}
        </tbody>
      </Table>
    </>
  );
};
const FlagsWithStatusIndicator = withStatusIndicator(FlagsContent);

FlagsContent.displayName = 'Flags';

const Flags: FC<RouteComponentProps> = () => {
  const pathPrefix = usePathPrefix();
  const { response, error, isLoading } = useFetch<FlagMap>(`${pathPrefix}/${API_PATH}/status/flags`);
  return <FlagsWithStatusIndicator data={response.data} error={error} isLoading={isLoading} />;
};

export default Flags;
