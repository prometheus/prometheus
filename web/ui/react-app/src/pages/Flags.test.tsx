import * as React from 'react';
import { shallow } from 'enzyme';
import { FlagsContent } from './Flags';
import { Table } from 'reactstrap';
import toJson from 'enzyme-to-json';

const sampleFlagsResponse = {
  'alertmanager.notification-queue-capacity': '10000',
  'alertmanager.timeout': '10s',
  'config.file': './documentation/examples/prometheus.yml',
  'log.format': 'logfmt',
  'log.level': 'info',
  'query.lookback-delta': '5m',
  'query.max-concurrency': '20',
  'query.max-samples': '50000000',
  'query.timeout': '2m',
  'rules.alert.for-grace-period': '10m',
  'rules.alert.for-outage-tolerance': '1h',
  'rules.alert.resend-delay': '1m',
  'storage.remote.flush-deadline': '1m',
  'storage.remote.read-concurrent-limit': '10',
  'storage.remote.read-max-bytes-in-frame': '1048576',
  'storage.remote.read-sample-limit': '50000000',
  'storage.tsdb.allow-overlapping-blocks': 'false',
  'storage.tsdb.max-block-duration': '36h',
  'storage.tsdb.min-block-duration': '2h',
  'storage.tsdb.no-lockfile': 'false',
  'storage.tsdb.path': 'data/',
  'storage.tsdb.retention': '0s',
  'storage.tsdb.retention.size': '0B',
  'storage.tsdb.retention.time': '0s',
  'storage.tsdb.wal-compression': 'false',
  'storage.tsdb.wal-segment-size': '0B',
  'web.console.libraries': 'console_libraries',
  'web.console.templates': 'consoles',
  'web.cors.origin': '.*',
  'web.enable-admin-api': 'false',
  'web.enable-lifecycle': 'false',
  'web.external-url': '',
  'web.listen-address': '0.0.0.0:9090',
  'web.max-connections': '512',
  'web.page-title': 'Prometheus Time Series Collection and Processing Server',
  'web.read-timeout': '5m',
  'web.route-prefix': '/',
  'web.user-assets': '',
};

describe('Flags', () => {
  it('renders a table with properly configured props', () => {
    const w = shallow(<FlagsContent data={sampleFlagsResponse} />);
    const table = w.find(Table);
    expect(table.props()).toMatchObject({
      bordered: true,
      size: 'sm',
      striped: true,
    });
  });
  it('should not fail if data is missing', () => {
    expect(shallow(<FlagsContent />)).toHaveLength(1);
  });
  it('should match snapshot', () => {
    const w = shallow(<FlagsContent data={sampleFlagsResponse} />);
    expect(toJson(w)).toMatchSnapshot();
  });
});
