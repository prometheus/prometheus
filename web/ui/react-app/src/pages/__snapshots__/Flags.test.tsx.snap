// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Flags should match snapshot 1`] = `
<Fragment>
  <h2>
    Command-Line Flags
  </h2>
  <Table
    bordered={true}
    responsiveTag="div"
    size="sm"
    striped={true}
    tag="table"
  >
    <tbody>
      <tr
        key="alertmanager.notification-queue-capacity"
      >
        <th>
          alertmanager.notification-queue-capacity
        </th>
        <td>
          10000
        </td>
      </tr>
      <tr
        key="alertmanager.timeout"
      >
        <th>
          alertmanager.timeout
        </th>
        <td>
          10s
        </td>
      </tr>
      <tr
        key="config.file"
      >
        <th>
          config.file
        </th>
        <td>
          ./documentation/examples/prometheus.yml
        </td>
      </tr>
      <tr
        key="log.format"
      >
        <th>
          log.format
        </th>
        <td>
          logfmt
        </td>
      </tr>
      <tr
        key="log.level"
      >
        <th>
          log.level
        </th>
        <td>
          info
        </td>
      </tr>
      <tr
        key="query.lookback-delta"
      >
        <th>
          query.lookback-delta
        </th>
        <td>
          5m
        </td>
      </tr>
      <tr
        key="query.max-concurrency"
      >
        <th>
          query.max-concurrency
        </th>
        <td>
          20
        </td>
      </tr>
      <tr
        key="query.max-samples"
      >
        <th>
          query.max-samples
        </th>
        <td>
          50000000
        </td>
      </tr>
      <tr
        key="query.timeout"
      >
        <th>
          query.timeout
        </th>
        <td>
          2m
        </td>
      </tr>
      <tr
        key="rules.alert.for-grace-period"
      >
        <th>
          rules.alert.for-grace-period
        </th>
        <td>
          10m
        </td>
      </tr>
      <tr
        key="rules.alert.for-outage-tolerance"
      >
        <th>
          rules.alert.for-outage-tolerance
        </th>
        <td>
          1h
        </td>
      </tr>
      <tr
        key="rules.alert.resend-delay"
      >
        <th>
          rules.alert.resend-delay
        </th>
        <td>
          1m
        </td>
      </tr>
      <tr
        key="storage.remote.flush-deadline"
      >
        <th>
          storage.remote.flush-deadline
        </th>
        <td>
          1m
        </td>
      </tr>
      <tr
        key="storage.remote.read-concurrent-limit"
      >
        <th>
          storage.remote.read-concurrent-limit
        </th>
        <td>
          10
        </td>
      </tr>
      <tr
        key="storage.remote.read-max-bytes-in-frame"
      >
        <th>
          storage.remote.read-max-bytes-in-frame
        </th>
        <td>
          1048576
        </td>
      </tr>
      <tr
        key="storage.remote.read-sample-limit"
      >
        <th>
          storage.remote.read-sample-limit
        </th>
        <td>
          50000000
        </td>
      </tr>
      <tr
        key="storage.tsdb.allow-overlapping-blocks"
      >
        <th>
          storage.tsdb.allow-overlapping-blocks
        </th>
        <td>
          false
        </td>
      </tr>
      <tr
        key="storage.tsdb.max-block-duration"
      >
        <th>
          storage.tsdb.max-block-duration
        </th>
        <td>
          36h
        </td>
      </tr>
      <tr
        key="storage.tsdb.min-block-duration"
      >
        <th>
          storage.tsdb.min-block-duration
        </th>
        <td>
          2h
        </td>
      </tr>
      <tr
        key="storage.tsdb.no-lockfile"
      >
        <th>
          storage.tsdb.no-lockfile
        </th>
        <td>
          false
        </td>
      </tr>
      <tr
        key="storage.tsdb.path"
      >
        <th>
          storage.tsdb.path
        </th>
        <td>
          data/
        </td>
      </tr>
      <tr
        key="storage.tsdb.retention"
      >
        <th>
          storage.tsdb.retention
        </th>
        <td>
          0s
        </td>
      </tr>
      <tr
        key="storage.tsdb.retention.size"
      >
        <th>
          storage.tsdb.retention.size
        </th>
        <td>
          0B
        </td>
      </tr>
      <tr
        key="storage.tsdb.retention.time"
      >
        <th>
          storage.tsdb.retention.time
        </th>
        <td>
          0s
        </td>
      </tr>
      <tr
        key="storage.tsdb.wal-compression"
      >
        <th>
          storage.tsdb.wal-compression
        </th>
        <td>
          false
        </td>
      </tr>
      <tr
        key="storage.tsdb.wal-segment-size"
      >
        <th>
          storage.tsdb.wal-segment-size
        </th>
        <td>
          0B
        </td>
      </tr>
      <tr
        key="web.console.libraries"
      >
        <th>
          web.console.libraries
        </th>
        <td>
          console_libraries
        </td>
      </tr>
      <tr
        key="web.console.templates"
      >
        <th>
          web.console.templates
        </th>
        <td>
          consoles
        </td>
      </tr>
      <tr
        key="web.cors.origin"
      >
        <th>
          web.cors.origin
        </th>
        <td>
          .*
        </td>
      </tr>
      <tr
        key="web.enable-admin-api"
      >
        <th>
          web.enable-admin-api
        </th>
        <td>
          false
        </td>
      </tr>
      <tr
        key="web.enable-lifecycle"
      >
        <th>
          web.enable-lifecycle
        </th>
        <td>
          false
        </td>
      </tr>
      <tr
        key="web.external-url"
      >
        <th>
          web.external-url
        </th>
        <td />
      </tr>
      <tr
        key="web.listen-address"
      >
        <th>
          web.listen-address
        </th>
        <td>
          0.0.0.0:9090
        </td>
      </tr>
      <tr
        key="web.max-connections"
      >
        <th>
          web.max-connections
        </th>
        <td>
          512
        </td>
      </tr>
      <tr
        key="web.page-title"
      >
        <th>
          web.page-title
        </th>
        <td>
          Prometheus Time Series Collection and Processing Server
        </td>
      </tr>
      <tr
        key="web.read-timeout"
      >
        <th>
          web.read-timeout
        </th>
        <td>
          5m
        </td>
      </tr>
      <tr
        key="web.route-prefix"
      >
        <th>
          web.route-prefix
        </th>
        <td>
          /
        </td>
      </tr>
      <tr
        key="web.user-assets"
      >
        <th>
          web.user-assets
        </th>
        <td />
      </tr>
    </tbody>
  </Table>
</Fragment>
`;
